### ISSUE TITLE
Arabic language (right to left in writing) stored (left to write) after create PDF Searchable

### ISSUE THREAD

**Issue by tbadran** [tesseract-ocr/tesseract#238]
I have tested latest release 3.05 on windows platform to OCR Arabic document to PDF (searchable) and when choose text from output PDF file it seems stored in opposite (left to write) and letters should be stored from (Right to left)!!!

i.e. original text I'm Arabic is مرحبا
Stored in PDF as text as ابحرم

**Comment by roozgar** [comment]
​please put your sample file and the command you used for ocr job​

**Comment by tbadran** [comment]
This is the command:

 

tesseract  c:\temp\test_ara.jpg  -l ara  -psm 3  c:\temp\test_ara pdf

 

Files are attached (source JPG and output PDF)

**Comment by tbadran** [comment]
Command and Samples are attached now in the previous comment

**Comment by amitdo** [comment]
Which program you are using to view the PDF?

**Comment by amitdo** [comment]
It does not look reversed wtth chrome pdf viewer, just not very accurate...

**Comment by roozgar** [comment]
@amitdo 
is there any way to reach a better accuracy in Arabic language until to change to new engine?
now with tesseract i get about 100% accuracy  in English but for Arabic result is about 30-40%
but for example i checked google drive ocr for Arabic and i see it have 100 results for same image..

can we work on language data for a better results?

**Comment by tbadran** [comment]
I am using Adobe Reader.
But please note that words are not reversed while viewing the PDF because it contains the original image with text layer.
I mean when you copy text layer then paste it to any text editor it will be reversed, so now can't search for the text inside the PDF because it is stored revered inside the text layer!

**Comment by tbadran** [comment]
This is a serious issue with the PDF output feature using Arabic Language and similar languages that be written from right to left

**Comment by amitdo** [comment]
cc: @jbreiden

**Comment by amitdo** [comment]
@roozgar

First, although I contributed some code and documentation to the project, I'm not an insider. 99% of the Tesseract code is written by devs from Google.

The current OCR engine used for Arabic is 'Cube'.

What Ray said/wrote about Cube?

> Hmm cube.
Its a dead-end. It didn't really contribute very much so nobody at google wants to work on it...  The *only* thing cube was really good at was Hindi. 

>Cube added Convolutional NN, but improvement was
disappointing.

>###Cube
>
* Standard Convolutional Neural Network character classifier.
* Uses over-segmentation and (a different) beam search for
character segmentation.
* Implemented as an alternative word recognizer.
* Achieves about 50% reduction in errors on Hindi when run together
with Tesseract’s word recognizer.
* Improvement for other languages disappointing: few % for major
slow-down (3x).
* Training code unmaintaine, unused.

>The next version is likely to include a new classifier implementation that uses neural nets. (bye bye cube probably.)

It seems that Ray is planning to release soon (in march?) a new (beta?) version of Tesseract (4.0?), that will include a new OCR engine based on LSTM.


What Ray said/wrote about LSTM?

>LSTM Integration
* LSTM code based on OCROpus Python implementation.
* Expanded capabilities including 2-D, and input of Tesseract
features, as well as images.
* Fully integrated with Tesseract at the line level. (May change)
* Visualization with existing Viewer API.
* Training code included (unlike cube).

With LSTM, OCR for printed Arabic (not real handwrite) can reach 95% character accuracy.

"Offline Printed Urdu Nastaleeq Script Recognition
with Bidirectional LSTM Networks"
http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.447.4577&rep=rep1&type=pdf

**Comment by amitdo** [comment]
> I checked google drive ocr for Arabic and i see it have 100 results for same image..

Neither you or I know what programs they are using to do OCR there...

Let's examine the logical options:
1) They only using Tesseract.
2) They using Tesseract, but their **private** version of it is more advanced than the  **public** open source current version.
3) They also doing extra image preprocessing and language model post-processing + some other 'tricks' to improve the final results.
4) They using Tesseract + other OCR programs in combination ('voting').
5) They are not using Tesseract at all.

1, 5 - Very unlikely options
2, 3 - Very likely
4 - Likely

**Comment by amitdo** [comment]
@tbadran

> But please note that words are not reversed while viewing the PDF because it contains the original image with text layer.
I mean when you copy text layer then paste it to any text editor it will be reversed, so now can't search for the text inside the PDF because it is stored revered inside the text layer!

Yes, I know...

Here is a copy of the invisible text layer (copied & pasted):

مداها ينم همهما
اللغة العريية
لغة جهد مه
مسنره هي انحاء العالم

Using Chromium (Google browser) PDF viewer under Linux.

**Comment by jbreiden** [comment]
I try hard to make sure Arabic and other right-to-left languages work correctly in Tesseract PDF. As the problem is isolated further I'm happy to look, but I'm not aware of any reason things would have broken.

**Comment by jbreiden** [comment]
A quick check shows Chrome gives good results (as per amitdo) and Acroread gives bad results (as per tbadran). This is surprising, I thought we were good with Acroread. I wonder if this is a regression and if so when it occurred.

**Comment by jbreiden** [comment]
Regarding recognition accuracy, that's a better topic for the forum. But in short: Don't compare against Google Drive. Don't expect major accuracy improvements unless/until Ray is successful with his ideas. And most importantly, don't trust any predictions about 'soon'. That's true for all software everywhere, and most definitely includes Tesseract.

**Comment by amitdo** [comment]
@roozgar 

You can try training Tesseract using the regular engine. Use the the wiki and see #169.I really don't know how good the result will be for Arabic.

Like jbreiden said, the timeline could change...

**Comment by tbadran** [comment]
Please note my testing using the binaries for Windows downloaded from:
http://domasofan.spdns.eu/tesseract/ 
and I am Using Windows 10 with Acrobat Pro 11 to view output PDF file

**Comment by tbadran** [comment]
I have tested multiple different sample files not only sample uploaded above and every time getting same issue in output PDF on windows 10 + Acrobat Pro 11

**Comment by tfmorris** [comment]
On OS X, I'm seeing the opposite of earlier reports: 
- Acrobat Reader DC 15.10.20056.167417 appears correct when cutting & pasting
- Google Chrome Version 48.0.2564.116 (64-bit) appears backwards

**Comment by tfmorris** [comment]
Adobe Acrobat:

امهمه مني اهادم
ةييرعلا ةغللا
. هم دهج ةغل
ملاعلا ءاحنا يه هرنسم

Google Chrome

مداها ينم همهما
اللغة العريية
لغة جهد مه
مسنره هي انحاء العالم

**Comment by amitdo** [comment]
@tom 
Look at the origininal jpg.
Lines 2 and 4 in Google Chrome look quite similar to lines 2 and 3 in the origininal jpg. lines 2 and 4 in lar First word in line 3 in the original jpg became first word in line 3 in Google Chrome.
Clearly, that's the 'good' output...

**Comment by amitdo** [comment]
Again, in Google Chromium.
If I mark the first two lines in the PDF + first word in line 3, 
copy the (invisible) text, paste it to a text file, 
mark the second to last word in line 3 in the PDF, 
copy the (invisible) text, paste it to the text file, I get:


مداها ينم همهما
اللغة العريية
لغة مسنره هي انحاء العالم

**Comment by jbreiden** [comment]
I find it a little easier to test with Hebrew because the letters do not connect. Tesseract version 3.03 behaves the same, so this is not a regression. Will need to think about this, because it is not obvious what exactly is going wrong. Lots of PDF files do a crazy 'write it backwards' strategy but that should not be required. Tesseract writes in reading order.

**Comment by jbreiden** [comment]
There are two things I can think of doing. One is to give up and write Arabic 
backwards (which I really hate!). The other is  to put an entry in  the PDF
metadata, Catalog/ViewerPreferences/Direction. Will continue thinking about
this, slowly.

**Comment by amitdo** [comment]
@jbreiden 
I didn't understand you. In one comment you talk about Hebrew and in another one you only referring Arabic. Does Hebrew displayed correctly with Adobe Reader?

**Comment by amitdo** [comment]
Please make sure that any change you do is not causing any regression with Chrome PDF viewer and OS X Preview. Thanks for your work!

**Comment by jbreiden** [comment]
@amitdo Hebrew has the exact same problem as Arabic.

**Comment by amitdo** [comment]
Maybe explicitly using unicode bidi control characters can help ?

**Comment by jbreiden** [comment]
That's another possibility, thanks for the suggestion.

**Comment by amitdo** [comment]
@jbreiden, any progress? Which way you chose?
Personally, I care about our Hebrew support.

**Comment by jbreiden** [comment]
I am taking a look at this today.

**Comment by jbreiden** [comment]
From the PDF 1.7 specification suggests using a left-to-right 
transformation matrix (Tm) while giving each character a negative 
width. A very crude experiment along these lines give good
results to Adobe Reader but messes up highlighting on Chrome.

```
In writing systems that are read from right to left (such as Arabic or Hebrew), 
one might expect that the glyphs in a font would have their origins at the lower right
and their widths (rightward horizontal displacements) specified as negative. 
```

```
--- tesseract/api/pdfrenderer.cpp	2016-07-06 13:19:57.000000000 -0700
+++ /tesseract/api/pdfrenderer.cpp	2016-07-06 15:35:12.000000000 -0700
@@ -246,6 +246,7 @@
 void AffineMatrix(int writing_direction,
                   int line_x1, int line_y1, int line_x2, int line_y2,
                   double *a, double *b, double *c, double *d) {
+  writing_direction = WRITING_DIRECTION_LEFT_TO_RIGHT;
   double theta = atan2(static_cast<double>(line_y1 - line_y2),
                        static_cast<double>(line_x2 - line_x1));
   *a = cos(theta);
@@ -527,7 +528,7 @@
                "endobj\n",
                5L,         // CIDToGIDMap
                7L,         // Font descriptor
-               1000 / kCharWidth);
+               - 1000 / kCharWidth);
   if (n >= sizeof(buf)) return false;
   AppendPDFObject(buf);
 ```

**Comment by amitdo** [comment]
@jbreiden
The PDF 1.7 spec refer to:
>Unicode Standard Annex #9, The Bidirectional Algorithm, Version **4.0.0**

http://www.unicode.org/reports/tr9/tr9-11.html

Support for RLI and PDI has been added in Unicode **6.3**.
http://www.unicode.org/reports/tr9/tr9-29.html

**Comment by jbreiden** [comment]
I tried the other control characters U+202b RIGHT-TO-LEFT EMBEDDING and U+202e RIGHT-TO-LEFT OVERRIDE. Neither had any effect with Adobe Reader 9.

```c++
--- tesseract/api/pdfrenderer.cpp	2016-07-06 13:19:57.000000000 -0700
+++ tesseract/api/pdfrenderer.cpp	2016-07-07 10:55:41.000000000 -0700
@@ -410,6 +410,9 @@
     bool last_word_in_block = res_it->IsAtFinalElement(RIL_BLOCK, RIL_WORD);
     STRING pdf_word("");
     int pdf_word_len = 0;
+    pdf_word += "<202E>";
+
+    pdf_word_len++;
     do {
       const char *grapheme = res_it->GetUTF8Text(RIL_SYMBOL);
       if (grapheme && grapheme[0] != '\0') {

```

[heb.pdf](https://github.com/tesseract-ocr/tesseract/files/352783/heb.pdf)

**Comment by jbreiden** [comment]
Filed feature request with Adobe to recognize -1 0 0 1 X Y Tm. No idea if they will consider it.

**Comment by Shreeshrii** [comment]
There are a number of issues relating to RTL and Arabic. Can they all be labelled with 'Arabic' for ease of finding, so that duplicate issues are not created.

https://github.com/tesseract-ocr/tesseract/issues?q=Arabic+in%3Atitle%2Cbody
gives a list of the same.

**Comment by amitdo** [comment]
Hi @Shreeshrii! 

Let see...

 #169
This is not Arabic specific issue, but an RTL issue. The reported issue was solved.

#212
A question, not an issue.

#238
PDF issue related to RTL. Not Arabic specific issue.

#294
'Moved' to tesseract-ocr/langdata issues reports.

#302
Seems to be solved.

#325
Original issue was solved.

#361
A broad complaint about bad RTL support.

#410
Not Arabic specific. Can't be solved.

As said before, once the new LSTM code will finally land in Tesseract's public Github repo, the OCR accuracy of Arabic and Persian will be dramatically improved. Cube code will be removed, so any issue with it will be irrelevant.

My conclusion: #238 is the only one in the list we should monitor. 

The big question left is when we will see Tesseract 4.0 code. Unfortunately, Ray does not yet share any planned date with the Tesseract community :(

**Comment by zdenop** [comment]
Ray shared that he would like to have public alpha version by the end of September.

**Comment by stweil** [comment]
That's good news. I promise that we'll give it a try as soon as it is available.

**Comment by amitdo** [comment]
@stweil,

>we'll give it a try...

We'? The @UB-Mannheim team I guess... :)

**Comment by Shreeshrii** [comment]
Thanks.

ShreeDevi
____________________________________________________________
भजन - कीर्तन - आरती @ http://bhajans.ramparivar.com

**Comment by jbreiden** [comment]
I'm currently in discussion with some Adobe folks about this topic.

**Comment by mehmetaltuntas** [comment]
hi, where can i get the arabic tessdata files? 
also, where do we get all other language files?
thanks

**Comment by Shreeshrii** [comment]
ara.* from https://github.com/tesseract-ocr/tessdata (Version 3.02)

https://github.com/tesseract-ocr/langdata/tree/master/ara (Version 3.04)

**Comment by Shreeshrii** [comment]
https://github.com/tesseract-ocr/tessdata

Download all ara.* Files for Arabic

Other language data files are also in same repository

**Comment by Shreeshrii** [comment]
The tesseract/langdata/ara repo has the 3.04 source files for Arabic
language data.

The Arabic traineddata is based on cube engine and is the 3.02version.

**Comment by amitdo** [comment]
@jbreiden
Did you find a solution?

**Comment by roozgar** [comment]
​is there any milestone to drop cube completely!?​

**Comment by amitdo** [comment]
>​is there any milestone to drop cube completely!?​

This issue is not caused by cube.

See https://github.com/tesseract-ocr/tesseract/issues/40#issuecomment-263039665

**Comment by jbreiden** [comment]
The Adobe folks suggested a few things to try, none of which worked so far. Still open and (relatively) active.

**Comment by jbreiden** [comment]
Okay, this bug has been open forever. As mentioned before, most PDF files deal with right-to-left (RTL) languages like Hebrew and Arabic by laying out the characters from left-to-right (LTR) but doing it backwards. This offends my programming sensibilities on many levels, and I've resisted this approach. But maybe it is time to swallow pride and wallow in the mud.  Here's a few examples from the test suite. How  is compatibility for search and copy-paste?

Arabic
[ara.pdf](https://github.com/tesseract-ocr/tesseract/files/947818/ara.pdf)

Single word Hebrew
[simplest.pdf](https://github.com/tesseract-ocr/tesseract/files/947819/simplest.pdf)

Hebrew + English
[heb_mivne.pdf](https://github.com/tesseract-ocr/tesseract/files/947854/heb_mivne.pdf)

Hebrew + English, tilted
[heb-tilt.pdf](https://github.com/tesseract-ocr/tesseract/files/947848/heb-tilt.pdf)

English (should be no change from what we do now)
[2.pdf](https://github.com/tesseract-ocr/tesseract/files/947838/2.pdf)

```c++
--- tesseract/api/pdfrenderer.cpp	2017-03-31 14:35:03.000000000 -0700
+++ tesseract/api/pdfrenderer.cpp	2017-04-21 10:16:23.000000000 -0700
@@ -225,14 +225,10 @@
 // left-to-right no matter what the reading order is. We need the
 // word baseline in reading order, so we do that conversion here. Returns
 // the word's baseline origin and length.
-void GetWordBaseline(int writing_direction, int ppi, int height,
+void GetWordBaseline(int ppi, int height,
                      int word_x1, int word_y1, int word_x2, int word_y2,
                      int line_x1, int line_y1, int line_x2, int line_y2,
                      double *x0, double *y0, double *length) {
-  if (writing_direction == WRITING_DIRECTION_RIGHT_TO_LEFT) {
-    Swap(&word_x1, &word_x2);
-    Swap(&word_y1, &word_y2);
-  }
   double word_length;
   double x, y;
   {
@@ -260,15 +256,12 @@
 }
 
 // Compute coefficients for an affine matrix describing the rotation
-// of the text. If the text is right-to-left such as Arabic or Hebrew,
-// we reflect over the Y-axis. This matrix will set the coordinate
+// of the text. This matrix will set the coordinate
 // system for placing text in the PDF file.
 //
-//                           RTL
-// [ x' ] = [ a b ][ x ] = [-1 0 ] [ cos sin ][ x ]
-// [ y' ]   [ c d ][ y ]   [ 0 1 ] [-sin cos ][ y ]
-void AffineMatrix(int writing_direction,
-                  int line_x1, int line_y1, int line_x2, int line_y2,
+// [ x' ] = [ a b ][ x ] = [ cos sin ][ x ]
+// [ y' ]   [ c d ][ y ]   [-sin cos ][ y ]
+void AffineMatrix(int line_x1, int line_y1, int line_x2, int line_y2,
                   double *a, double *b, double *c, double *d) {
   double theta = atan2(static_cast<double>(line_y1 - line_y2),
                        static_cast<double>(line_x2 - line_x1));
@@ -276,17 +269,6 @@
   *b = sin(theta);
   *c = -sin(theta);
   *d = cos(theta);
-  switch(writing_direction) {
-    case WRITING_DIRECTION_RIGHT_TO_LEFT:
-      *a = -*a;
-      *b = -*b;
-      break;
-    case WRITING_DIRECTION_TOP_TO_BOTTOM:
-      // TODO(jbreiden) Consider using the vertical PDF writing mode.
-      break;
-    default:
-      break;
-  }
 }
 
 // There are some really awkward PDF viewers in the wild, such as
@@ -407,15 +389,14 @@
     {
       int word_x1, word_y1, word_x2, word_y2;
       res_it->Baseline(RIL_WORD, &word_x1, &word_y1, &word_x2, &word_y2);
-      GetWordBaseline(writing_direction, ppi, height,
+      GetWordBaseline(ppi, height,
                       word_x1, word_y1, word_x2, word_y2,
                       line_x1, line_y1, line_x2, line_y2,
                       &x, &y, &word_length);
     }
 
     if (writing_direction != old_writing_direction || new_block) {
-      AffineMatrix(writing_direction,
-                   line_x1, line_y1, line_x2, line_y2, &a, &b, &c, &d);
+      AffineMatrix(line_x1, line_y1, line_x2, line_y2, &a, &b, &c, &d);
       pdf_str.add_str_double(" ", prec(a));  // . This affine matrix
       pdf_str.add_str_double(" ", prec(b));  // . sets the coordinate
       pdf_str.add_str_double(" ", prec(c));  // . system for all
@@ -459,23 +440,34 @@
     bool last_word_in_block = res_it->IsAtFinalElement(RIL_BLOCK, RIL_WORD);
     STRING pdf_word("");
     int pdf_word_len = 0;
+    GenericVector<int> unicodes;
+
+    // Gather up unicode codepoints for the word
     do {
       const char *grapheme = res_it->GetUTF8Text(RIL_SYMBOL);
       if (grapheme && grapheme[0] != '\0') {
-        GenericVector<int> unicodes;
         UNICHAR::UTF8ToUnicode(grapheme, &unicodes);
-        char utf16[kMaxBytesPerCodepoint];
-        for (int i = 0; i < unicodes.length(); i++) {
-          int code = unicodes[i];
-          if (CodepointToUtf16be(code, utf16)) {
-            pdf_word += utf16;
-            pdf_word_len++;
-          }
-        }
       }
       delete []grapheme;
       res_it->Next(RIL_SYMBOL);
     } while (!res_it->Empty(RIL_BLOCK) && !res_it->IsAtBeginningOf(RIL_WORD));
+
+
+    // Use primitive "write it backwards" approach for RTL languages
+    if (writing_direction == WRITING_DIRECTION_RIGHT_TO_LEFT) {
+      unicodes.reverse();
+    }
+
+    // Write out the word the way PDF likes it
+    char utf16[kMaxBytesPerCodepoint];
+    for (int i = 0; i < unicodes.length(); i++) {
+      int codepoint = unicodes[i];
+      if (CodepointToUtf16be(codepoint, utf16)) {
+        pdf_word += utf16;
+        pdf_word_len++;
+      }
+    }
+
     if (word_length > 0 && pdf_word_len > 0 && fontsize > 0) {
       double h_stretch =
           kCharWidth * prec(100.0 * word_length / (fontsize * pdf_word_len));

```

**Comment by amitdo** [comment]
Please tell us which pdf viewer you already tested if any.

**Comment by jbreiden** [comment]
So far, just Pdfium/Chrome/Linux   AdobeReader9/Linux  Preivew/MacOSX.  I've asked a few different groups of people to take a look.

**Comment by jbreiden** [comment]
@christophered Can you please take a look at ara.pdf above and tell me if it works well for you?

**Comment by christophered** [comment]
I had a quick look at ara.pdf via Chrome and Windows PDF Reader and it seems fine, though there are some errors, such as الانسان is showing as االنسان , the  الا errors. and some mistakes here and there.

**Comment by christophered** [comment]
How did you train Tesseract 4  for Arabic?

**Comment by jbreiden** [comment]
@christophered Thanks, that's a very helpful report. I expect it to be equal to or better than what Tesseact does now for producing PDF of Arabic.  Note that my change is just about PDF generation and does not touch the recognization process in any way. It is just stock Tesseract 4.0  support for Arabic.

What Tesseract does right now:
[control.pdf](https://github.com/tesseract-ocr/tesseract/files/953433/control.pdf)

**Comment by christophered** [comment]
@jbreiden  Here is a more thorough examination of "ara.pdf" [that you posted in your comment](https://github.com/tesseract-ocr/tesseract/issues/238#issuecomment-296256374)

1) Wrongly sentenced ( Sentence disorder ): 
    some sentences are over-stepping their location in the paragraph.

![ara](https://cloud.githubusercontent.com/assets/16248376/25373915/1eb574ba-29a4-11e7-8c5d-5bb4d0d1480d.jpg)


2) Repetitive mistakes:
    ( لا ) is wrongly represented as ( ال ) , which is actually opposite to the correct spelling. 
    ( ، ) is the Arabic Comma, is wrongly represented as ( ء ) or ( , ) or ( . ) or ( » )
    ( اً ) is represented by only ( ا ) , which is missing (  ً   ) 
    some rare cases of multiple combined words, there are 2 separate cases ( مرحامستبشرا ) and( منالناس ) ,    should be ( مرحاً مستبشرا ) and ( من الناس )


3) Rare Case:
    When I copied the text to Microsoft Word, most of the font was in Arial font except of a couple of ( . ) full-stops which were in Calibri font, a weird thing to see.

Conclusion:
Altogether, except for the mistakes that I stated earlier, the recognition rate was very good in this sample.

Note that in the Arabic language, the state of ( لا ) is frequently used, so-much that if this misrepresentation of it as ( ال ) is continued, it would degrade the recognition rate drastically.

**Comment by amitdo** [comment]
Hebrew report:

I highlighted the text in each pdf viewer, and pasted it to gedit.

With Chromium the straight version is mostly fine. There are problems when there is a combination of Hebrew and English/other ltr symbols in the same line.
The skewed version is not fine. The words appear in wrong order in each line.

The pasted text of the straight version does not look good when Evince and pdf.js are used.
pdf.js - there are line breaks after each word.
Evince - total mess. Wrong line breaks and wrong word order. Unusable.

**Comment by amitdo** [comment]
It will be helpful to compare pasted text of these files to Tesseract's text renderer output, to see if each issue is really caused by the pdf renderer or by the ocr engine itself.

**Comment by amitdo** [comment]
I think we should also compare to the current (4.00 with lstm) pdf output, without your patch.
The original issue, reversed letters, was with the Adobe pdf viewer, not the other viewers.

**Comment by amitdo** [comment]
@christophered

The source of the mistakes in your 'Repetitive mistakes' section is the ocr engine itself, not the pdf renderer.

#648 is a more suitable place to report about them.
( لا ) vs ( ال ) is a known issue. See https://github.com/tesseract-ocr/tesseract/issues/648#issuecomment-285633162 and the comments below it.

**Comment by amitdo** [comment]
>some rare cases of multiple combined words

It's not clear if the source for this issue is the pdf renderer or the ocr engine itself.

**Comment by christophered** [comment]
I discovered that the Sentence disorder is caused by Chrome which I used to view the PDF.
Note that after using Windows Reader the PDF was viewed correctly without the disorder that I mentioned.

**Comment by amitdo** [comment]
@christophered,
We want that it will be displayed fine in all major pdf viewers, not just one.

**Comment by amitdo** [comment]
>Note that after using Windows Reader the PDF was viewed correctly

Which Windows version?

If you have Windows 10, try to open the pdf file with the Edge browser, and report how it is displayed there.

**Comment by christophered** [comment]
I am using Windows 8

**Comment by Shreeshrii** [comment]
Amit, the PDF displays the original image only, so lookswise it will be the
same. It is the text layer, as copied or saved which is different.

I can test on windows10 and post the result. Someone else will have to tell
if it is ok or not.

Even with legacy Devanagari fonts that use Latin range, I have found that
copied text is different between Adobe reader and foxit reader.

- excuse the brevity, sent from mobile

**Comment by amitdo** [comment]
Me:
>If you have Windows 10, try to open the pdf file with the Edge browser, and report how it is displayed there.

@Shreeshrii :
>Amit, the PDF displays the original image only, so lookswise it will be the same. It is the text layer, as copied or saved which is different. 

Yes. although I said 'displayed', I meant to refer to the "invisible text layer underneath the visible image, which can be copied and pasted to text editor".

I just made a shortcut, assuming people will understand what I really meant. 
Maybe I should add **excuse the brevity, sent from my PC** to my comments...
:rofl:

**Comment by jbreiden** [comment]
Thank you for the reports, and sorry for any and all confusion. I'm going to post what Tesseract does right now (CONTROL) and also repost the proposed modification (EXPERIMENT). I am especially interested in regressions, anywhere EXPERIMENT does worse than CONTROL.

CONTROL
[2.pdf](https://github.com/tesseract-ocr/tesseract/files/956615/2.pdf)
[ara.pdf](https://github.com/tesseract-ocr/tesseract/files/956613/ara.pdf)
[heb_mivne.pdf](https://github.com/tesseract-ocr/tesseract/files/956614/heb_mivne.pdf)
[heb-tilt.pdf](https://github.com/tesseract-ocr/tesseract/files/956612/heb-tilt.pdf)
[simplest.pdf](https://github.com/tesseract-ocr/tesseract/files/956616/simplest.pdf)


EXPERIMENT  (repost)
[2.pdf](https://github.com/tesseract-ocr/tesseract/files/956604/2.pdf)
[ara.pdf](https://github.com/tesseract-ocr/tesseract/files/956605/ara.pdf)
[heb_mivne.pdf](https://github.com/tesseract-ocr/tesseract/files/956606/heb_mivne.pdf)
[heb-tilt.pdf](https://github.com/tesseract-ocr/tesseract/files/956608/heb-tilt.pdf)
[simplest.pdf](https://github.com/tesseract-ocr/tesseract/files/956607/simplest.pdf)

**Comment by christophered** [comment]
@amitdo started to get funny  :)

**Comment by Shreeshrii** [comment]
I downloaded ara.pdf (EXPERIMENT) version and opened in Adobe Reader XI and Foxit Reader 8.1 under Windows 10 and copied and pasted the text in Notepad++ under Windows 10.

Both text files look different - with text order being opposite of each other. Please see attached.

![ara-error](https://cloud.githubusercontent.com/assets/5095331/25424593/72b755be-2a86-11e7-8f27-411ddd9bc0f5.png)

[ARA-ADOBE-XI.txt](https://github.com/tesseract-ocr/tesseract/files/957656/ARA-ADOBE-XI.txt)

[ARA-FOXIT-8.1.txt](https://github.com/tesseract-ocr/tesseract/files/957657/ARA-FOXIT-8.1.txt)

**Comment by Shreeshrii** [comment]
Additionally, on windows 10, in both Edge as well as the windows 10 pdf reader I am not able to copy text. 

Text copied in Chrome version is similar to Foxit, but with additional line breaks. @christophered can check whether the order is also changed.

![ara-chrome-error](https://cloud.githubusercontent.com/assets/5095331/25425054/42b71aa0-2a88-11e7-9891-e55475dd16b7.png)
[ara-chrome.txt](https://github.com/tesseract-ocr/tesseract/files/957672/ara-chrome.txt)

**Comment by Shreeshrii** [comment]
Windows 10 - Internet Explorer - text matches output from Adobe Reader XI.

[ara-internet-explorer.txt](https://github.com/tesseract-ocr/tesseract/files/957692/ara-internet-explorer.txt)

**Comment by christophered** [comment]
@Shreeshrii ARA-FOXIT-8.1.txt is the most adequate one in terms of sentence organization.

**Comment by amitdo** [comment]
I only checked the straight version of the Hebrew document.

The control is much worst with Evince and pdf.js. The letters themselves are reversed.

Chrome: The two look the same, except the two last line (RTL).
Both are not good, but the experiment is worst than the control the digits in the zip code and the phone arw nunbers are in the wrong order.

I will check the skewed version later.

**Comment by amitdo** [comment]
Skewed version, Chromium:

Both have wrong word order in each (but not exactly the same).

Experiment:
Last lines - same issue as the straight version.
The second word in the document is missing.
Two separate lines become one. Hebrew line + English line (site address).

**Comment by amitdo** [comment]
@jbreiden , any update about this issue? Should Zdenko apply your patch?

**Comment by christophered** [comment]
@amitdo 
Using the latest Tesseract  4.0 alpha and the latest best Arabic model, I created a searchable pdf output:

- When using Chrome to view the pdf, the text can be selected/copied/pasted correctly (RTL).
- When using Adobe Acrobat Reader 17.012 (latest to date), though the text is displayed correctly, but when selected/pasted, is in reverse text (LTR).

**Comment by christophered** [comment]
@amitdo 
Using the latest ABBYY FineReader 14 to create a searchable pdf:
- Both Chrome and Adobe Acrobat Reader can select/copy/paste correctly.

Conclusion:
It seems that Tesseract needs tweaking to solve this problem.

[Original Image.zip](https://github.com/tesseract-ocr/tesseract/files/1257071/Original.Image.zip)
[Tesseract.pdf](https://github.com/tesseract-ocr/tesseract/files/1257075/Tesseract.pdf)
[Abby Finereader.pdf](https://github.com/tesseract-ocr/tesseract/files/1257074/Abby.Finereader.pdf)

**Comment by jbreiden** [comment]
Looking at this again. Slowly losing the remainder of my sanity.

**Comment by amitdo** [comment]
>It seems that Tesseract needs tweaking to solve this problem.

The patch was not applied yet, so the original issue still exist.

>Looking at this again. Slowly losing the remainder of my sanity.

It could be worst if you were rapidly losing your sanity :-)

You are using simple reverse here. That's not good enough for bidi text.

**Comment by msaleh83** [comment]
Any luck fixing this issue yet?!

**Comment by amitdo** [comment]
@jbreiden 

Should we apply your patch?

**Comment by jbreiden** [comment]
Don't apply the patch, because it didn't work very well.

**Comment by Shreeshrii** [comment]
wikipedia says:

```
If Tesseract is used to process right-to-left text such as Arabic or Hebrew, the results are ordered as though it is left-to-right text.[11]
```

Is this still true for current version?

**Comment by jbreiden** [comment]
PDF output in Tesseract is and has always been reading order.

**Comment by Shreeshrii** [comment]
@jbreiden So, should I delete that line on the wikipedia page?

**Comment by godmanalafandy** [comment]
Hi guys I found the solution and I hope you visit my channel on YouTube and you will 
https://www.youtube.com/channel/UCHAGGT45tNfMkhJ9i2wQeFQ?view_as=subscriber
find the video tomorrow
1. convert pdf to word
2. convert language from en to ar
3. convert word to pdf
That all thing >>>>> thank for you

**Comment by godmanalafandy** [comment]
Hi guys I found the solution and I hope you visit my channel on YouTube and you will 
https://www.youtube.com/channel/UCHAGGT45tNfMkhJ9i2wQeFQ?view_as=subscriber
find the video tomorrow
1. convert pdf to word
2. convert language from en to ar
3. convert word to pdf
That all thing >>>>> thank for you

**Comment by godmanalafandy** [comment]
This solution and you should see for yourself ☺☺☺
https://www.youtube.com/watch?v=JXXOKBdCdE0

**Comment by stweil** [comment]
What is the status of this issue? Is there still a problem which should be fixed for Tesseract 4.0.0?

**Comment by jbreiden** [comment]
Status is still unsolved, and currently inactive. I swallowed my pride and
experimented
with writing arabic backwards in PDF like everyone else does, and it still
didn't work nicely.

**Comment by jbreiden** [comment]
Correction: ** currently inactive **. I still think it is important though.

**Comment by stweil** [comment]
This sounds as if there will not be a fix in the near future. So we should not require that this bug must be fixed for 4.0.0.

**Comment by MalekBadi** [comment]
Is this issue resolved ??

**Comment by yregaieg** [comment]
> @amitdo
> Using the latest ABBYY FineReader 14 to create a searchable pdf:
> 
>     * Both Chrome and Adobe Acrobat Reader can select/copy/paste correctly.
> 
> 
> Conclusion:
> It seems that Tesseract needs tweaking to solve this problem.
> 
> [Original Image.zip](https://github.com/tesseract-ocr/tesseract/files/1257071/Original.Image.zip)
> [Tesseract.pdf](https://github.com/tesseract-ocr/tesseract/files/1257075/Tesseract.pdf)
> [Abby Finereader.pdf](https://github.com/tesseract-ocr/tesseract/files/1257074/Abby.Finereader.pdf)

@jbreiden I have experimented with the files he attached, and I came to notice something that does actually make sense : It seems both files have different setup for text orientation when I start selecting from mid-sentence and drag over a few lines :
Tesseract :
<img width="501" alt="image" src="https://user-images.githubusercontent.com/11628365/58635949-7304cd80-82ef-11e9-8a9a-8d9ea2b803b9.png">
ABBY :
<img width="572" alt="image" src="https://user-images.githubusercontent.com/11628365/58635971-82841680-82ef-11e9-8a2f-92c2de67ebdb.png">

**Comment by yregaieg** [comment]
Not sure if this is of any value : https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdfs/PDF32000_2008.pdf 
Page 598 (606 / 756) has a table describing writing mode :
<img width="799" alt="image" src="https://user-images.githubusercontent.com/11628365/58637693-59fe1b80-82f3-11e9-9bdc-02180de7ed75.png">
And it seems to be possible to specify direction of writing as RTL using this parameter on a structure element and all child elements. @jbreiden can you have a look at it ?

Section 14.8.2.3.3 could be related to this issue as well.

**Comment by ReactNativeFan** [comment]
I would like to inform that the problem still persists in **Tesseract 4.1.1**
The Tesseract recognizes and displays Arabic text correctly. However, when export results as PDF/A, the stored text in PDF/A are reversed. 

Yes, if you open the PDF in Acrobat, it will give you reversed words, and will work fine for Google Chrome PDF reader. However, when i extracted the stored text in PDF/A using pdfToText, the words are reversed too, which means the text was stored in the wrong order.

See the following example for more details:

![ar](https://user-images.githubusercontent.com/45165854/84205238-522e9680-aa72-11ea-8c25-4c688866824a.png)

Here is the PDF/A generated by Tesseract
[Recognized_PDFA_By_Tesseract.pdf](https://github.com/tesseract-ocr/tesseract/files/4755080/Recognized_PDFA_By_Tesseract.pdf)

To summarize:

**True Text**
مرحبا بكم جميعا
اللغة العربية

**Tesseract Text** 100% correct
مرحبا بكم جميعا
اللغة العربية

**Tesseract PDF/A Text**  
اعيمج مكب ابحرم 
ةيبرعلا ةغللا      

As you see in the Tesseract PDF/A text, every word is reversed although the .hOCR file is correct.

![hOCR](https://user-images.githubusercontent.com/45165854/84206347-9327aa80-aa74-11ea-9154-cee852b7dd4b.png)

**Comment by Mennaruuk** [comment]
This issue also persists in Tesseract 5 alpha. Another issue is when I double-click to select then copy a word, my computer does copy correctly the whole word. Visibly, however, the blue selection box isn't going over the entire word. You can observe that in the screenshot below. Both of these issues occur in Adobe Reader DC and Okular.
![image](https://user-images.githubusercontent.com/52135169/105098320-e4226700-5a77-11eb-992a-b38c380c559f.png)

**Comment by diyajunaid** [comment]
Please advise if this issue is resolved in any latest version of tesseract?

**Comment by saleha-DS** [comment]
I'm using Tess4j 4.5.1 and having the same issue. When I process image and create pdf, open it in Acrobat reader, it displays 100% correct. However, search doesn't work unless I reverse the letters. When I copy the text and paste it in MS Word, it display reversed.

**Comment by amitdo** [comment]
Unfortunately, this long standing issue was not solved.

Copy&paste or search of words in documents written in RTL scripts only works with Google Chrome's PDF reader., Even with this viewer there might be some issues.

**Comment by wollmers** [comment]
> Unfortunately, this long standing issue was not solved.
> 
> Copy&paste or search of words in documents written in RTL scripts only works with Google Chrome's PDF reader., Even with this viewer there might be some issues.

RTL scripts are stored left to right (storage order) according to standards, but only rendered and displayed RTL.

In the above example it's stored the right way, but the github editor can not handle it correctly. But I can use a screenshot of my command line:

![Bildschirmfoto 2022-02-21 um 23 06 16](https://user-images.githubusercontent.com/1275557/155031660-29193f05-4446-432f-9a49-728ebc8fd01b.png)

Also on my MacOS the program Preview displays the PDF correctly and it is searchable with copy & paste into the search field:

![Bildschirmfoto 2022-02-21 um 23 00 02](https://user-images.githubusercontent.com/1275557/155031806-942aac8c-3cb0-482e-bf79-147f5e6c0895.png)

Thus it's a problem of PDF viewers, not of Tesseract.

**Comment by amitdo** [comment]
>RTL scripts are stored left to right (storage order) according to standards, but only rendered and displayed RTL.

Which standards are you referring here?

Here is how it works in html:
https://www.w3.org/International/questions/qa-visual-vs-logical.

Regarding the PDF standard:

https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdfs/PDF32000_2008.pdf

Tesseract does not do what is described in: 
**14.8.2.3.3 Reverse-Order Show Strings**

**Comment by wollmers** [comment]
> > RTL scripts are stored left to right (storage order) according to standards, but only rendered and displayed RTL.
> 
> Which standards are you referring here?

Unicode
https://unicode.org/reports/tr9/#Introduction

> Here is how it works in html: https://www.w3.org/International/questions/qa-visual-vs-logical.

HTML just implements the Unicode standard. And all editors, command line clients I know use logical storage order.

> Regarding the PDF standard:
> 
> https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdfs/PDF32000_2008.pdf
> 
> Tesseract does not do what is described in: **14.8.2.3.3 Reverse-Order Show Strings**

That's an obscurity of the PDF specification. Yes, in the above PDF sample I didn't find the string `/ReversedChars`. Does this mean, Tesseract should implement this storage method? 

At least Acrobat Reader should do it right for the interfaces to the "normal" world like search field. Other PDF viewers do it.

**Comment by amitdo** [comment]
It not just Adobe, Firefox and Evince also has the same issue.

Years ago, there was an attempt to implement the strings reversal as described in the spec, to make Adobe and other viewers happy, but it didn't work so well.

**Comment by amitdo** [comment]
There is a regression with Google's Chrome. It **completely** fail to render the Arabic pdf above. Same failure occur with another pdf with Hebrew.

```
$ chromium --version
Chromium 98.0.4758.102 snap
```

**Comment by florisre** [comment]
As of now (using tesseract 5.3.1), the issue still exists. Brave (Chromium-based) handles my Arabic-script PDF correctly, Firefox and Okular do not. Haven't tested anything else.

**Comment by IrtazaIjaz** [comment]
Hi @tbadran/ @amitdo/ @ryanfb,

My Urdu text is following below which I have used for training:
**نقش فریادی ہے کس کی شوخیٔ تحریر کا**

![page_10_line_1](https://github.com/tesseract-ocr/tesseract/assets/29355799/c8241507-7c87-4e9c-96f2-42e834301b67)

I'm using the above image to perform get the accurate output. However, after training it is returning below:

**اتریشےکہک کا یہ ےیج ایاےک**

Please help me with what to do to perform the exact outcome.

Attached is my page_10_line_1.box file made by the below command:

```
# /content/tesstrain Path to run the below code
!make training MODEL_NAME=irt START_MODEL=urd FINETUNE_TYPE=Impact
```

[page_10_line_1.zip](https://github.com/tesseract-ocr/tesseract/files/13209812/page_10_line_1.zip)
